A67-xds-dynamic-parameters.md

```diff
@@ -0,0 +1,246 @@
+           "channel_creds": [ { "type": "google_default" } ]
+         }
+       ],
+>      "dynamic_parameters": {
+>        "env": "prod"
+>      }
+     }
+   },
+   "node": {
+     "id": "projects/123456789123/networks/default/nodes/4d5f8bd2-3e91-4940-855a-fdabca4c2f70"
+   }
+ }
+```
+
+To support non-xdstp use cases, we can add a map at the root level too:
+
+```text
+ {
+   "xds_servers": [
+     {
+       "server_uri": "xds.example.com:443",
+       "channel_creds": [ { "type": "google_default" } ]
+       ],
+       "server_features": [
+         "xds_v3"
+       ]
+     }
+   ],
+   "node": {
+     "id": "projects/123456789123/networks/default/nodes/4d5f8bd2-3e91-4940-855a-fdabca4c2f70"
+   },
+>  "dynamic_parameters": {
+>    "env": "prod"
+>  }
+ }
+```
+
+The `XdsClient` is currently structured as follows in all languages:
+
+-   There is a bootstrap config known by the `XdsClient` object. The bootstrap
+    config contains settings for each authority.
+-   We have a shared pool of xDS channel objects, so if multiple authorities use
+    the same xDS server, they will share the same xDS channel object.
+-   The code that constructs the ADS request already knows the set of resource
+    names to send, and that list may include resources from multiple authorities
+    that are sharing the same xDS server.
+
+The changes this design proposes are:
+
+-   The bootstrap config will contain the dynamic parameters to use for each
+    authority.
+-   The code to construct the ADS request will be changed such that for each
+    authority for which there are resource names in the request, it looks up the
+    dynamic parameters for that authority in the bootstrap config. If the
+    dynamic parameters are not empty, that will trigger using the
+    `resource_locators` field instead of the `resource_names` field in the
+    request
+-   The code to parse the ADS response will be changed to look in the
+    `resource_name` field and use this instead of the name field if it exists.
+    As this `XdsClient` is only a leaf client, and not a caching proxy, we can
+    ignore the dynamic_parameter_constraints field of the ResourceName message.
+
+The updated code to construct the ADS request would look something like this
+(pseudo code):
+
+```text
+// Example authority_resource_names:
+[
+  // Resources for authority "foo.example.com"
+  { resource_names: ["bar", "baz"], dynamic_parameters: {"env": "prod"} },
+
+  // Resources for authority "zzz.example.com"
+  { resource_names: ["yyy"], dynamic_parameters: {"env": "prod", "version": "v2"} },
+]
+
+function ConstructADSRequest(authority_resource_names)
+  ...
+  foreach resource_names, dynamic_parameters in authority_resource_names
+    if dynamic_parameters is empty
+      request.resource_names = resource_names
+    else
+      foreach name in resource_names
+        request.resource_locators.add(name, dynamic_parameters)
+  ...
+```
+
+### Temporary environment variable protection
```

Comment:
@markdroth You mentioned out of band that you think we do need a flag, so I changed it to say so. (This actually almost takes away the reason for not being able to reuse the metadata.)

You also mentioned that we will need an interop test. Should I mention this, or is that just a given, perhaps? Grep'ing for "interop" in the proposals repo, it doesn't seem like this is something that is usually explicitly mentioned.

Reply:
Yeah, we definitely want env var protection here.

I don't think this actually eliminates the reason for not using the metadata.  In practice, I think we're going to want to explictly control when users upgrade to dynamic parameters, not just based on whether they're using a version of gRPC that has passed interop tests.

We definitely need an interop test definition for interop with TD, but that's Google-specific, not part of this OSS gRFC, so I don't think it needs to be included in this doc.  Please take that as a separate action item.